connector/failoverconnector/failover.go

```diff
@@ -12,11 +17,216 @@ type consumerProvider[C any] func(...component.ID) (C, error)
 type failoverRouter[C any] struct {
 	consumerProvider consumerProvider[C]
 	cfg              *Config
+	pS               *pipelineSelector
+	consumers        []C
+	cancelRetry      *cancelManager
+}
+
+// Manages cancel function for retry goroutine, ends up cleaner than using channels
+type cancelManager struct {
+	lock        sync.Mutex
+	cancelRetry context.CancelFunc
+}
+
+func (m *cancelManager) updateCancelFunc(newCancelFunc context.CancelFunc) {
+	m.lock.Lock()
+	m.cancelRetry = newCancelFunc
+	m.lock.Unlock()
+}
+
+func (m *cancelManager) invokeCancel() {
+	m.lock.Lock()
+	if m.cancelRetry != nil {
+		m.cancelRetry()
+	}
+	m.lock.Unlock()
 }
 
+var (
+	errNoValidPipeline = errors.New("All provided pipelines return errors")
+	errConsumer        = errors.New("Error registering consumer")
+)
+
 func newFailoverRouter[C any](provider consumerProvider[C], cfg *Config) *failoverRouter[C] {
 	return &failoverRouter[C]{
 		consumerProvider: provider,
 		cfg:              cfg,
+		pS:               newPipelineSelector(cfg),
+		cancelRetry:      &cancelManager{},
+	}
+}
+
+func (f *failoverRouter[C]) getCurrentConsumer() C {
+	return f.consumers[f.pS.currentIndex]
+}
+
+func (f *failoverRouter[C]) registerConsumers() error {
+	consumers := make([]C, 0)
+	for _, pipeline := range f.cfg.PipelinePriority {
+		newConsumer, err := f.consumerProvider(pipeline...)
+		if err == nil {
+			consumers = append(consumers, newConsumer)
+		} else {
+			return errConsumer
+		}
+	}
+	f.consumers = consumers
+	return nil
+}
+
+// handlePipelineError evaluates if the error was due to a failed retry or a new failure and will resolve the error
+// by adjusting the priority level accordingly
+func (f *failoverRouter[C]) handlePipelineError() {
+	if f.pS.handleErrorRetryCheck() {
+		ctx, cancel := context.WithCancel(context.Background())
+		f.cancelRetry.invokeCancel()
+		f.cancelRetry.updateCancelFunc(cancel)
+		f.enableRetry(ctx)
+	}
+}
+
+func (f *failoverRouter[C]) enableRetry(ctx context.Context) {
+
+	var cancelFunc context.CancelFunc
+	var tickerCtx context.Context
+	ticker := time.NewTicker(f.cfg.RetryInterval)
```

Comment:
Why keep these outside of the goroutine?

Reply:
No real reason, can be moved